src/Orleans.Core/Async/TaskExtensions.cs

```diff
@@ -1,13 +1,135 @@
+        public static Task<object> Box(this Task task)
+        {
+            switch (task.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    return CompletedTask;
+
+                case TaskStatus.Faulted:
+                    return TaskFromFaulted(task);
+
+                case TaskStatus.Canceled:
+                    return CanceledTask;
+
+                default:
+                    return BoxAwait(task);
+            }
+        }
+
+        /// <summary>
+        /// Returns a <see cref="Task{Object}"/> for the provided <see cref="Task{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">The underlying type of <paramref name="task"/>.</typeparam>
+        /// <param name="task">The task.</param>
+        public static Task<object> Box<T>(this Task<T> task)
+        {
+            if (typeof(T) == typeof(object))
+                return task as Task<object>;
+
+            switch (task.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    return Task.FromResult((object)GetResult(task));
+
+                case TaskStatus.Faulted:
+                    return TaskFromFaulted(task);
+
+                case TaskStatus.Canceled:
+                    return CanceledTask;
+
+                default:
+                    return BoxAwait(task);
+            }
+        }
+
+        /// <summary>
+        /// Returns a <see cref="Task{Object}"/> for the provided <see cref="Task{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">The underlying type of <paramref name="task"/>.</typeparam>
+        /// <param name="task">The task.</param>
+        internal static Task<T> Unbox<T>(this Task<object> task)
+        {
+            if (typeof(T) == typeof(object))
+                return task as Task<T>;
+
+            switch (task.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    return Task.FromResult((T)GetResult(task));
+
+                case TaskStatus.Faulted:
+                    return TaskFromFaulted<T>(task);
+
+                case TaskStatus.Canceled:
+                    return TaskFromCanceled<T>();
+
+                default:
+                    return UnboxContinuation<T>(task);
+            }
+        }
+
+        /// <summary>
+        /// Returns a <see cref="Task{Object}"/> for the provided <see cref="Task{Object}"/>.
+        /// </summary>
+        /// <param name="task">The task.</param>
+        public static Task<object> Box(this Task<object> task)
+        {
+            return task;
+        }
+
+        private static async Task<object> BoxAwait(Task task)
+        {
+            await task;
+            return default(object);
+        }
+
+        private static async Task<object> BoxAwait<T>(Task<T> task)
+        {
+            return await task;
+        }
+
+        private static Task<T> UnboxContinuation<T>(Task<object> task)
+        {
```

Comment:
I believe there's room for improvement here and in related methods, but we can do that in another PR.

Reply:
Absolutely, we can also update codegen to avoid so much boxing, at least separating the paths for Task and TaskOfT